File: packages/sdk-python/agentlens/redactor.py
"""PII redactor — same patterns as the TypeScript SDK.

Each pattern is replaced with a bracketed label so consumers can see
that a value existed without retaining the sensitive data.

Usage::

    from agentlens.redactor import redact, maybe_redact

    clean = redact("Contact user@example.com or call 415-555-1234")
    # -> "Contact [REDACTED-EMAIL] or call [REDACTED-PHONE]"
"""

from __future__ import annotations

import re

# ---------------------------------------------------------------------------
# Redaction rules — identical patterns to the TypeScript SDK's redactor.ts
# ---------------------------------------------------------------------------

_RULES: list[tuple[str, re.Pattern[str]]] = [
    (
        "EMAIL",
        # RFC 5321-ish — intentionally broad for PII detection
        re.compile(r"\b[A-Za-z0-9._%+\-]+@[A-Za-z0-9.\-]+\.[A-Za-z]{2,}\b"),
    ),
    (
        "PHONE",
        # North-American + international E.164 variants
        re.compile(r"(?:\+?1[\s.\-]?)?\(?\d{3}\)?[\s.\-]?\d{3}[\s.\-]?\d{4}\b"),
    ),
    (
        "SSN",
        re.compile(r"\b(?!000|666|9\d{2})\d{3}[- ]?\d{2}[- ]?\d{4}\b"),
    ),
    (
        "CREDIT_CARD",
        # Luhn-formatted 13-19 digit card numbers with optional separators
        re.compile(r"\b(?:\d[ \-]?){13,19}\b"),
    ),
    (
        "API_KEY",
        # Common API key patterns: sk-…, pk-…, key-…, token-… (32+ chars)
        re.compile(
            r"\b(?:sk|pk|key|api_key|token|secret)[_\-]?[A-Za-z0-9]{20,}\b",
            re.IGNORECASE,
        ),
    ),
    (
        "BEARER_TOKEN",
        # Authorization: Bearer <token>
        re.compile(r"\bBearer\s+[A-Za-z0-9\-._~+/]+=*\b", re.IGNORECASE),
    ),
]


def redact(text: str) -> str:
    """Scan *text* for PII patterns and replace each match with
    ``[REDACTED-<TYPE>]``.  Returns the sanitised string.
    """
    result = text
    for label, pattern in _RULES:
        result = pattern.sub(f"[REDACTED-{label}]", result)
    return result


def maybe_redact(text: str | None, enabled: bool) -> str | None:
    """Conditionally redact *text*; returns ``None`` / ``undefined`` unchanged."""
    if not enabled or text is None:
        return text
    return redact(text)

File: packages/sdk-core/src/redactor.ts
/**
 * Lightweight PII redactor applied to `input` and `output` text before
 * spans are buffered. Enabled via `AgentLensConfig.redactPII = true`.
 *
 * Each pattern is replaced with a bracketed label so consumers can see that
 * a value existed without retaining the sensitive data.
 */

interface RedactionRule {
  label: string;
  pattern: RegExp;
}

const RULES: RedactionRule[] = [
  {
    label: 'EMAIL',
    // RFC 5321-ish — intentionally broad for PII detection
    pattern: /\b[A-Za-z0-9._%+\-]+@[A-Za-z0-9.\-]+\.[A-Za-z]{2,}\b/g,
  },
  {
    label: 'PHONE',
    // North-American + international E.164 variants
    pattern: /(?:\+?1[\s.-]?)?\(?\d{3}\)?[\s.-]?\d{3}[\s.-]?\d{4}\b/g,
  },
  {
    label: 'SSN',
    pattern: /\b(?!000|666|9\d{2})\d{3}[- ]?\d{2}[- ]?\d{4}\b/g,
  },
  {
    label: 'CREDIT_CARD',
    // Luhn-formatted 13-19 digit card numbers with optional separators
    pattern: /\b(?:\d[ -]?){13,19}\b/g,
  },
  {
    label: 'API_KEY',
    // Common API key patterns: sk-…, pk-…, key-…, token-… (32+ chars)
    pattern: /\b(?:sk|pk|key|api_key|token|secret)[_-]?[A-Za-z0-9]{20,}\b/gi,
  },
  {
    label: 'BEARER_TOKEN',
    // Authorization: Bearer <token>
    pattern: /\bBearer\s+[A-Za-z0-9\-._~+/]+=*\b/gi,
  },
];

const REPLACEMENT = (label: string): string => `[REDACTED-${label}]`;

/**
 * Scans `text` for PII patterns and replaces each match with
 * `[REDACTED-<TYPE>]`.  Returns the sanitised string.
 */
export function redact(text: string): string {
  let result = text;
  for (const rule of RULES) {
    // Reset lastIndex for global regexes between calls
    rule.pattern.lastIndex = 0;
    result = result.replace(rule.pattern, REPLACEMENT(rule.label));
  }
  return result;
}

/**
 * Conditionally redacts `text`. Returns `undefined` unchanged.
 */
export function maybeRedact(text: string | undefined, enabled: boolean): string | undefined {
  if (!enabled || text === undefined) return text;
  return redact(text);
}
